### DISCUSSION TITLE
PermissionDenied Error when using with j4rs

### DISCUSSION THREAD

**Discussion by wangxiaoying** [PyO3/pyo3#2315] [Questions] [answered]
I'm using [j4rs](https://github.com/astonbitecode/j4rs) to run java code by invoking a jar file. The code works fine when I run the code using rust. However, it will throw the following error when I invoke the same function from python:

```
pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: GeneralError("Error { kind: PermissionDenied, message: \"Operation not permitted (os error 1)\" }")
```

 The code panicked at the line of code for [adding the jar file to class path](https://github.com/astonbitecode/j4rs#adding-jars-to-the-classpath).

Here is the permission of the jar file:
```
ll my.jar
-rwxrwxrwx  1 username  everyone    24M 14 Apr 23:16 my.jar
```

Any idea what's going on and how to fix it? Please let me know if you need more information!

Here is a repo that could reproduce the error: https://github.com/wangxiaoying/pyo3-j4rs-example

Both `cargo run --example test` and `maturin develop && python test.py` will invoke the [run_java](https://github.com/wangxiaoying/pyo3-j4rs-example/blob/main/src/run_java.rs) function, while the first successes and the later throws the error.

**Comment by birkenfeld** [comment]
I think we can't say much about this without a reproducing example, or at least the full code you're running. For example, in the link you gave there is no `unwrap`.

**Reply by wangxiaoying** [reply]
Thanks for the quick response! Here is a repo with minimum code that could reproduce the result: https://github.com/wangxiaoying/pyo3-j4rs-example

**Comment by wangxiaoying** [comment]
Here is a repo with minimum code that could reproduce the result: https://github.com/wangxiaoying/pyo3-j4rs-example.

Both `cargo run --example test` and `maturin develop && python test.py` will invoke the [run_java](https://github.com/wangxiaoying/pyo3-j4rs-example/blob/main/src/run_java.rs) function, while the first successes and the later throws the error.

**Comment by birkenfeld** [comment, marked as answer]
Thank you! I'm getting a different error
```
pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: GeneralError("Error { kind: NotFound, message: \"No such file or directory (os error 2)\" }")
```
but the cause is probably the same:

j4rs tries to find the `jassets` path which is under `target/debug/jassets`. It tries to find that in the same directory (or up) of the current executable. For Rust-only, this is under `target/debug/...` which works. For Python, it is `/usr/bin/python`, and there is no "jassets" there.

You need to set the proper path with https://docs.rs/j4rs/latest/j4rs/struct.JvmBuilder.html#method.with_base_path - this worked for me:
```
    let base = fs::canonicalize("./target/debug").unwrap();
    let jvm: Jvm = JvmBuilder::new().classpath_entry(entry).with_base_path(base.to_str().unwrap()).build().unwrap();
```
but of course the path will be different for different targets, and after deployment, so you probably need to add another env. var for it.

**Reply by wangxiaoying** [reply]
Thank you so much for such a quick response. Setting the base path solves the problem perfectly!
